src/librustc_trait_selection/traits/error_reporting/suggestions.rs

```diff
@@ -1424,96 +1412,122 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
             format!("does not implement `{}`", trait_ref.print_only_trait_path())
         };
 
-        if let Some(await_span) = from_awaited_ty {
-            // The type causing this obligation is one being awaited at await_span.
-            let mut span = MultiSpan::from_span(await_span);
+        if let Some((target_span, scope_span, yield_span, expr, from_awaited_ty)) = interior {
```

Comment:
I wish we could find a way to reduce the nesting here. It might be time to factor this out into a helper function.